### DISCUSSION TITLE
Printing string(varchar) to serial output

### DISCUSSION THREAD

**Discussion by electro-cr** [tdslite/tdslite#5] [Q&A] [answered]
So I got this up and running on a ESP32, looks great! but when I try to get a text result from the database and print to the console I got some extra characters at print out.

This is my callback to print column 0 (int) and 1(varchar(50) on database):

static void row_callback(void * u, const tdsl::tds_colmetadata_token & colmd,
                         const tdsl::tdsl_row & row) {
    SERIAL_PRINTLNF("row: %d %s", row [0].as<tdsl::int32_t>(),
                    row [1].as<tdsl::char_view>().data());
}

And print out is:
Executing query: SELECT * FROM product
row: 1 CABLEUSBA����
row: 2 7271-1735-3W
row: 3 7271-1822-3W
row: 14 CABLE_B
Rows affected: 4

How can I modify the callback to print correctly? Why is getting more data than the actual data on the database for that column?

Thanks!

**Comment by xmkg** [comment, marked as answer]
Hello @electro-cr,

The problem here is character fields in TDS (e.g. varchar, nvarchar) are not NUL-terminated, and "char_view" is just a "view" to that (i.e. no copy, no modification). SERIAL_PRINTLNF macro uses snprintf in the background and it expects a NUL terminator to stop printing characters for the "%s" format specifier and continues to print the consequent characters until one is found.

Given that the varchar fields you're trying to print lack a \0 at the end, you'd need to use the `%.*s` format specifier to let snprintf know how many characters to print (assuming esp's snprintf implementation supports that):

```
SERIAL_PRINTLNF("row: %d %.*s" row[0].as<tdsl::uint32_t>(), row[1].as<tdsl::char_view>().size(), row[1].as<tdsl::char_view>().data());
```

I know that Arduino's stdlib does not support "%.*s", but esp's might. Let's see whether that works before moving to the alternatives.

Cheers,

**Reply by electro-cr** [reply]
Hi @mustafakemalgilor, thanks for the answare and explanation of what's going on. Look like this method is supported by arduino ide for the ESP32, only had to correctly assign the rows:

`SERIAL_PRINTLNF("row: %d %.*s", row[0].as<tdsl::int32_t>(), row[1].as<tdsl::char_view>().size(), row[1].as<tdsl::char_view>().data());`

Output from print:
Executing query: SELECT * FROM product
row: 1 CABLEUSBA
row: 2 7271-1735-3W
row: 3 7271-1822-3W
row: 14 CABLE_B
Rows affected: 4

Cheers!

**Comment by xmkg** [comment]
Glad to hear that it worked :)
